src/execution_plan/execution_plan.c

```diff
@@ -498,304 +482,360 @@ static void _NewExecutionPlanSegment(RedisModuleCtx *ctx, GraphContext *gc,
-			AR_ExpNode *exp = projections[i];
-			AR_EXP_CollectEntityIDs(exp, modifies_ids);
-		}
+	uint skip = 0;
+	uint limit = RESULTSET_UNLIMITED;
+	if(skip_clause) skip = AST_ParseIntegerNode(skip_clause);
+	if(limit_clause) limit = AST_ParseIntegerNode(limit_clause);
 
-		modifies = array_new(uint, raxSize(modifies_ids));
-		raxIterator iter;
-		raxStart(&iter, modifies_ids);
-		raxSeek(&iter, ">=", (unsigned char *)"", 0);
-		while(raxNext(&iter)) {
-			modifies = array_append(modifies, *(uint *)iter.key);
-		}
-		raxFree(modifies_ids);
+	int sort_direction = 0;
+	if(segment->order_expressions) {
+		const cypher_astnode_t *order_clause = cypher_ast_with_get_order_by(clause);
+		sort_direction = AST_PrepareSortOp(order_clause);
 	}
 
-	OpBase *op;
+	_buildProjectionOps(segment, skip, limit, sort_direction, aggregate, distinct, ops);
+}
 
-	if(with_clause) {
-		// uint *with_projections = AST_WithClauseModifies(ast, with_clause);
-		if(AST_ClauseContainsAggregation(with_clause)) {
-			op = NewAggregateOp(projections, modifies);
-		} else {
-			op = NewProjectOp(projections, modifies);
-		}
-		Vector_Push(ops, op);
+static void _ExecutionPlanSegment_ConvertClause(GraphContext *gc, AST *ast,
+												ExecutionPlanSegment *segment, ResultSetStatistics *stats, const cypher_astnode_t *clause,
+												Vector *ops) {
+	OpBase *op = NULL;
+	cypher_astnode_type_t t = cypher_astnode_type(clause);
+	// Because 't' is set using the offsetof() call, it cannot be used in switch statements.
+	if(t == CYPHER_AST_MATCH || t == CYPHER_AST_CALL) {
+		// MATCH and CALL clauses are processed before looping over all clauses.
+		// TODO This should be changed later
+		return;
+	} else if(t == CYPHER_AST_CREATE) {
+		// Only add at most one Create op per segment. TODO Revisit and improve this logic.
+		if(_ExecutionPlan_ContainsCreateOp(ops)) return;
+		AST_CreateContext create_ast_ctx = AST_PrepareCreateOp(gc, segment->record_map, ast,
+															   segment->query_graph);
+		op = NewCreateOp(stats, create_ast_ctx.nodes_to_create, create_ast_ctx.edges_to_create);
+	} else if(t == CYPHER_AST_UNWIND) {
+		AST_UnwindContext unwind_ast_ctx = AST_PrepareUnwindOp(clause, segment->record_map);
+		op = NewUnwindOp(unwind_ast_ctx.record_idx, unwind_ast_ctx.exp);
+	} else if(t == CYPHER_AST_MERGE) {
+		// A merge clause provides a single path that must exist or be created.
+		// As with paths in a MATCH query, build the appropriate traversal operations
+		// and append them to the set of ops.
+		AST_MergeContext merge_ast_ctx = AST_PrepareMergeOp(gc, segment->record_map, ast, clause,
+															segment->query_graph);
+		op = NewMergeOp(stats, merge_ast_ctx.nodes_to_merge, merge_ast_ctx.edges_to_merge);
+	} else if(t == CYPHER_AST_SET) {
+		uint nitems;
+		EntityUpdateEvalCtx *update_exps = AST_PrepareUpdateOp(clause, segment->record_map, &nitems);
+		op = NewUpdateOp(gc, update_exps, nitems, stats);
+	} else if(t == CYPHER_AST_DELETE) {
+		uint *nodes_ref;
```

Comment:
In my eyes it would be cleaner for each if condition to call a single function similar to }
```else if(t == CYPHER_AST_RETURN) {
		// Converting a RETURN clause can create multiple operations.
		_buildReturnOps(segment, clause, ops);
```

Reply:
:+1: